src/application/tester.go

```diff
@@ -26,45 +28,60 @@ func TestApp(appContext types.AppContext, logger *util.Logger, mode composebuild
 		if serviceContext.Config.Development.Scripts["test"] == "" {
 			logger.Logf("%s has no tests, skipping", serviceContext.Dir)
 		} else {
-			testPassed, err := TestService(serviceContext, logger, mode)
-			if err != nil {
+			testResult, err := TestService(serviceContext, logger, mode, shutdown)
+			switch {
+			case testResult.Interrupted:
+				return testResult, err
+			case err != nil:
 				logger.Logf("error running '%s' tests:", err)
-			}
-			if !testPassed {
+				fallthrough
+			case !testResult.Passed:
 				numFailed++
 			}
 		}
 	}
 	if numFailed == 0 {
 		logger.Log("All tests passed")
-		return true, nil
+		return types.TestResult{
+			Passed: true,
+		}, nil
 	}
 	logger.Logf("%d tests failed", numFailed)
-	return false, nil
+	return types.TestResult{}, nil
 }
 
-// TestService runs the tests for the service and return true if the tests passed
-// and an error if any
-func TestService(serviceContext types.ServiceContext, logger *util.Logger, mode composebuilder.BuildMode) (bool, error) {
+// TestService runs the tests for the service and returns a TestResult struct
+func TestService(serviceContext types.ServiceContext, logger *util.Logger, mode composebuilder.BuildMode, shutdown chan os.Signal) (types.TestResult, error) {
 	logger.Logf("Testing service '%s'", serviceContext.Dir)
 	serviceTester, err := NewServiceTester(serviceContext, logger, mode)
 	if err != nil {
-		return false, err
+		return types.TestResult{}, err
 	}
 
-	exitCode, err := serviceTester.Run()
-	if err != nil {
-		return false, err
-	}
-	var testPassed bool
-	var result string
-	if exitCode == 0 {
-		testPassed = true
-		result = "passed"
-	} else {
-		testPassed = false
-		result = "failed"
+	testExit := make(chan int)
+	testError := make(chan error)
+	go func() {
+		exitCode, err := serviceTester.Start()
+		if err != nil {
+			testError <- err
+			return
+		}
+		testExit <- exitCode
+	}()
+
+	select {
+	case <-shutdown:
+		return types.TestResult{Interrupted: true}, serviceTester.Shutdown()
+	case err := <-testError:
+		serviceTester.Shutdown() // nolint errcheck
```

Comment:
I don't think we need to shutdown in this case and don't believe we were before either.